src/HydraAuctionOnchain/MintingPolicies/Auction.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module HydraAuctionOnchain.MintingPolicies.Auction
+  ( auctionMintingPolicy
+  ) where
+
+import HydraAuctionOnchain.Errors.MintingPolicies.Auction (PAuctionMpError (..))
+import HydraAuctionOnchain.Helpers (pdecodeInlineDatum, pfindUniqueOutputWithScriptHash)
+import HydraAuctionOnchain.Lib.ScriptContext (pownCurrencySymbol)
+import HydraAuctionOnchain.Types.AuctionInfo (PAuctionInfo)
+import HydraAuctionOnchain.Types.AuctionTerms (pvalidateAuctionTerms)
+import HydraAuctionOnchain.Types.Error (errCode, passert, passertMaybe)
+import HydraAuctionOnchain.Types.Tokens
+  ( pauctionTokenBundleBurned
+  , pauctionTokenBundleMinted
+  , ptxOutContainsAuctionMetadataToken
+  )
+import Plutarch.Api.V1.Value (pnormalize)
+import Plutarch.Api.V2 (PCurrencySymbol, PScriptContext, PScriptHash, PTxOutRef)
+import Plutarch.Extra.Maybe (pisJust)
+import Plutarch.Extra.ScriptContext (pfindTxInByTxOutRef)
+import Plutarch.Monadic qualified as P
+
+data PAuctionMpRedeemer (s :: S)
+  = MintAuctionRedeemer (Term s (PDataRecord '[]))
+  | BurnAuctionRedeemer (Term s (PDataRecord '[]))
+  deriving stock (Generic)
+  deriving anyclass (PlutusType, PIsData, PShow, PEq)
+
+instance DerivePlutusType PAuctionMpRedeemer where
+  type DPTStrat _ = PlutusTypeData
+
+auctionMintingPolicy
+  :: Term
+      s
+      ( PScriptHash
+          :--> PTxOutRef
+          :--> PAuctionMpRedeemer
+          :--> PScriptContext
+          :--> PUnit
+      )
+auctionMintingPolicy = phoistAcyclic $
+  plam $ \auctionMetadataSh nonceOref redeemer ctx -> P.do
+    -- Script purpose should contain own currency symbol. Always
+    -- true for minting policies.
+    auctionCs <-
+      plet $
+        passertMaybe
+          $(errCode AuctionMp'Error'MissingOwnCurrencySymbol)
+          (pownCurrencySymbol # ctx)
+
+    pmatch redeemer $ \case
+      MintAuctionRedeemer _ ->
+        pcheckMintAuction # auctionMetadataSh # nonceOref # ctx # auctionCs
+      BurnAuctionRedeemer _ ->
+        pcheckBurnAuction # ctx # auctionCs
+
+pcheckMintAuction
+  :: Term
+      s
+      ( PScriptHash
+          :--> PTxOutRef
+          :--> PScriptContext
+          :--> PCurrencySymbol
+          :--> PUnit
+      )
+pcheckMintAuction = phoistAcyclic $
+  plam $ \auctionMetadataSh nonceOref ctx auctionCs -> P.do
+    txInfo <- plet $ pfield @"txInfo" # ctx
+    txInfoFields <- pletFields @["mint", "inputs"] txInfo
+    mintValue <- plet $ pnormalize # txInfoFields.mint
+    let inputs = txInfoFields.inputs
+
+    -- The auction state, auction metadata, and standing bid tokens
+    -- of the auction should all be minted.
+    -- No other tokens should be minted or burned.
+    passert $(errCode AuctionMp'MintAuction'Error'AuctionTokenBundleNotMinted) $
+      mintValue #== pauctionTokenBundleMinted # auctionCs
```

Comment:
I realized I've never used to check multiple token names. Do we have any guarantees regarding tokens' being listed in a particular order or does #== care about that?

Reply:
This should be safe, since this equality check is only possible between `PValue` with sorted keys, which is annotated at the type level via `KeyGuarantees`.